diff -r -c asterisk-1.6.0.28.orig/channels/chan_sip.c asterisk-1.6.0.28/channels/chan_sip.c
*** asterisk-1.6.0.28.orig/channels/chan_sip.c	2010-05-11 15:35:37.000000000 -0400
--- asterisk-1.6.0.28/channels/chan_sip.c	2013-04-02 10:25:19.000000000 -0400
***************
*** 1982,1988 ****
  static void set_socket_transport(struct sip_socket *socket, int transport);
  
  /* Realtime device support */
! static void realtime_update_peer(const char *peername, struct sockaddr_in *sin, const char *username, const char *fullcontact, int expirey, int deprecated_username, int lastms);
  static struct sip_user *realtime_user(const char *username);
  static void update_peer(struct sip_peer *p, int expiry);
  static struct ast_variable *get_insecure_variable_from_config(struct ast_config *config);
--- 1982,1988 ----
  static void set_socket_transport(struct sip_socket *socket, int transport);
  
  /* Realtime device support */
! static void realtime_update_peer(const char *peername, struct sockaddr_in *sin, const char *username, const char *fullcontact, const char *useragent, int expirey, int deprecated_username, int lastms);
  static struct sip_user *realtime_user(const char *username);
  static void update_peer(struct sip_peer *p, int expiry);
  static struct ast_variable *get_insecure_variable_from_config(struct ast_config *config);
***************
*** 3690,3696 ****
  	that name and store that in the "regserver" field in the sippeers
  	table to facilitate multi-server setups.
  */
! static void realtime_update_peer(const char *peername, struct sockaddr_in *sin, const char *defaultuser, const char *fullcontact, int expirey, int deprecated_username, int lastms)
  {
  	char port[10];
  	char ipaddr[INET_ADDRSTRLEN];
--- 3690,3696 ----
  	that name and store that in the "regserver" field in the sippeers
  	table to facilitate multi-server setups.
  */
! static void realtime_update_peer(const char *peername, struct sockaddr_in *sin, const char *defaultuser, const char *fullcontact, const char *useragent, int expirey, int deprecated_username, int lastms)
  {
  	char port[10];
  	char ipaddr[INET_ADDRSTRLEN];
***************
*** 3721,3731 ****
  
  	if (fc)
  		ast_update_realtime(tablename, "name", peername, "ipaddr", ipaddr,
! 			"port", port, "regseconds", regseconds,
  			deprecated_username ? "username" : "defaultuser", defaultuser, fc, fullcontact, syslabel, sysname, NULL); /* note fc and syslabel _can_ be NULL */
  	else
  		ast_update_realtime(tablename, "name", peername, "ipaddr", ipaddr,
! 			"port", port, "regseconds", regseconds,
  			deprecated_username ? "username" : "defaultuser", defaultuser, syslabel, sysname, NULL); /* note syslabel _can_ be NULL */
  	/* We cannot do this in the same statement as above, because the lack of
  	 * this field could cause the whole statement to fail. */
--- 3721,3731 ----
  
  	if (fc)
  		ast_update_realtime(tablename, "name", peername, "ipaddr", ipaddr,
! 			"port", port, "regseconds", regseconds, "useragent", useragent,
  			deprecated_username ? "username" : "defaultuser", defaultuser, fc, fullcontact, syslabel, sysname, NULL); /* note fc and syslabel _can_ be NULL */
  	else
  		ast_update_realtime(tablename, "name", peername, "ipaddr", ipaddr,
! 			"port", port, "regseconds", regseconds, "useragent", useragent,
  			deprecated_username ? "username" : "defaultuser", defaultuser, syslabel, sysname, NULL); /* note syslabel _can_ be NULL */
  	/* We cannot do this in the same statement as above, because the lack of
  	 * this field could cause the whole statement to fail. */
***************
*** 3852,3858 ****
  	int rtcachefriends = ast_test_flag(&p->flags[1], SIP_PAGE2_RTCACHEFRIENDS);
  	if (sip_cfg.peer_rtupdate &&
  	    (p->is_realtime || rtcachefriends)) {
! 		realtime_update_peer(p->name, &p->addr, p->username, rtcachefriends ? p->fullcontact : NULL, expiry, p->deprecated_username, p->lastms);
  	}
  }
  
--- 3852,3858 ----
  	int rtcachefriends = ast_test_flag(&p->flags[1], SIP_PAGE2_RTCACHEFRIENDS);
  	if (sip_cfg.peer_rtupdate &&
  	    (p->is_realtime || rtcachefriends)) {
! 		realtime_update_peer(p->name, &p->addr, p->username, rtcachefriends ? p->fullcontact : NULL, p->useragent, expiry, p->deprecated_username, p->lastms);
  	}
  }
  
***************
*** 22181,22186 ****
--- 22181,22188 ----
  			sscanf(v->value, "%30d", &peer->lastms);
  		} else if (realtime && !strcasecmp(v->name, "ipaddr") && !ast_strlen_zero(v->value) ) {
  			inet_aton(v->value, &(peer->addr.sin_addr));
+ 		} else if (realtime && !strcasecmp(v->name, "useragent")) {
+ 			ast_copy_string(peer->useragent, v->value, sizeof(peer->useragent));
  		} else if (realtime && !strcasecmp(v->name, "fullcontact")) {
  			if (alt_fullcontact && !alt) {
  				/* Reset, because the alternate also has a fullcontact and we
